mithril-aggregator/src/file_uploaders/gcp_uploader/gcloud_backend.rs

```diff
@@ -0,0 +1,153 @@
+pub struct GCloudBackendUploader {
+    bucket: String,
+    use_cdn_domain: bool,
+    storage_client: GcpStorageClient,
+    logger: Logger,
+}
+
+impl GCloudBackendUploader {
+    /// Creates a new instance of `GCloudBackendUploader`
+    pub async fn try_new(
+        bucket: String,
+        use_cdn_domain: bool,
+        credentials_json_env_var: String,
+        logger: Logger,
+    ) -> StdResult<Self> {
+        const BASE_ERROR_CONTEXT: &str =
+            "Failed to create Google Cloud Storage client for file uploading";
+        let unparsed_credentials_json = std::env::var(&credentials_json_env_var)
+            .with_context(|| {
+                format!("Environment variable `{credentials_json_env_var}` must be set")
+            })
+            .with_context(|| BASE_ERROR_CONTEXT)?;
+        let credentials_file = CredentialsFile::new_from_str(&unparsed_credentials_json)
+            .await
+            .with_context(|| BASE_ERROR_CONTEXT)?;
+        let config = ClientConfig::default()
+            .with_credentials(credentials_file)
+            .await
+            .with_context(|| BASE_ERROR_CONTEXT)?;
+
+        Ok(Self {
+            bucket,
+            use_cdn_domain,
+            storage_client: GcpStorageClient::new(config),
+            logger,
+        })
+    }
+}
+
+#[async_trait]
+impl CloudBackendUploader for GCloudBackendUploader {
+    async fn file_exists(&self, remote_file_path: &CloudRemotePath) -> StdResult<Option<FileUri>> {
+        info!(self.logger, "Reading file metadata {remote_file_path}");
+        let request = GetObjectRequest {
+            bucket: self.bucket.clone(),
+            object: remote_file_path.to_string(),
+            ..Default::default()
+        };
+
+        let file_uri = match self
+            .storage_client
+            .get_object(&request)
+            .await
+            .with_context(|| "remote reading file metadata failure")
+        {
+            Ok(_) => {
+                info!(self.logger, "Found file metadata {remote_file_path}");
+
+                Some(remote_file_path.to_gcloud_storage_location(&self.bucket, self.use_cdn_domain))
+            }
+            Err(_) => {
+                info!(self.logger, "Missing file metadata {remote_file_path}");
+
+                None
+            }
+        };
+
+        Ok(file_uri)
+    }
+
+    async fn upload_file(
+        &self,
+        local_file_path: &Path,
+        remote_file_path: &CloudRemotePath,
+    ) -> StdResult<FileUri> {
+        info!(
+            self.logger,
+            "Uploading {} to {remote_file_path}",
+            local_file_path.display()
+        );
+        let file = tokio::fs::File::open(local_file_path)
+            .await
+            .with_context(|| {
+                format!("Opening file failure, path: {}", local_file_path.display())
+            })?;
+        let stream = FramedRead::new(file, BytesCodec::new());
+
+        self.storage_client
+            .upload_streamed_object(
```

Comment:
Are we sure that this method will overwrite an existing file?

Reply:
It use the following api (same as before): https://cloud.google.com/storage/docs/json_api/v1/objects/insert
This link specifies that:
> Stores a new object and metadata. The uploaded object replaces any existing object with the same name.
So it should overwrite existing files.